design/ceph-device-set.md

```diff
@@ -0,0 +1,86 @@
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: "failure-domain.beta.kubernetes.io/zone"
+            operator: In
+            values:
+            - us-west-1a
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "ocs.openshift.io/cluster"
+              operator: In
+              values:
+              - my-ceph-cluster
+          topologyKey: "failure-domain.beta.kubernetes.io/zone"
+```
+
+The CephDeviceSet CRD will define characteristics of a set of OSD devices and their Pods. The following fields are available in the spec:
+
+* **cluster**: The Ceph cluster the devices will belong to. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **capacity**: The capacity of the devices in this set. **[required]**
+* **storageClassName**: The StorageClass to provision the PVCs from. If not specified, the cluster default StorageClass will be used.
+* **resources**: The CPU and RAM requests/limits to apply to the OSD Pods. Default is no resource requests.
+* **placement**: The placement criteria to apply to the OSD Pods. Default is no placement criteria.
+
+## CephCluster CRD Requirements
+The CephCluster CRD will have to be extended to include a new field: `spec.storage.useDeviceSets`. If this is `true`, the CephCluster controller will skip the usual task of creating OsdPrepare Jobs and OSD Deployments. Configuring either `useAllDevices` or specific devices on nodes will be mutually exclusive from this new configuration option. Default will be `false`, to preserve current behavior.
+
+## Workflow
+To start, the admin will need to create a CephCluster CR with `spec.storage.useDeviceSets` set to `true`. This will cause the CephCluster controller to not discover, prepare, or create OSDs. The admin will then create one or more CephDeviceSet CRs to specify the storage for the cluster.
+
+If the admin wishes to control device placement beyond simply using node hostnames it will be up to them to make sure the desired nodes are labeled properly to ensure the Kuberenetes scheduler will distribute the OSD Pods appropriately.
+
+Additionally, it is recommended that the admin setup a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
+
+### OSDs as StatefulSets
+In this design, a set of OSDs will be represented by a StatefulSet with the following configuration:
+
+* `count` will map directly to `replicas` in the StatefulSet.
+* `resources` and `placement` will correspond directly to the `resources` and `affinity` fields for the StatefulSet Pod template
+* `capacity` and `storageClassName` will be used similarly in the StatefulSet's `volumeClaimTemplates` field.
+* The current OsdPrepare Job will be specified as an init container for the StatefulSet.
+
+### CephDeviceSet Controller
```

Comment:
After pondering on the separate controller for the deviceset the last couple days, I'm really seeing that a separate controller does not fit current patterns and we should go with the approach to add this to the existing `storage` element. The existing cephcluster CR controller should be used to manage the device sets.
- All storage nodes are currently managed by the `storage` element. 
- From a UX perspective, it makes less sense to define some OSDs in one CR and some in another CR.
- If device sets were handled by a separate controller, it would require coordination between the controllers, which adds a level of complexity to the implementation.
- A `name` field can be added to the device sets and enforced to be unique by the operator. This is logically equivalent to requiring a unique name when they are separate CRs.
- If in the same CR, the device sets can inherit other properties from the cephcluster CR. For example, the resource limits and requests and placement for OSDs can already be defined there. The device set could still override them, but they could at least follow the pattern of inheriting the settings.
- A new setting for `useDeviceSets` wouldn't be necessary. If device sets are defined, the other settings could just be ignored as needed. On the other hand, why wouldn't we allow current OSDs and the device set OSDs to co-exist? Some configurations may not make sense, but the CR gives the ability to express desired state. If the admin wants both to exist, they should both exist according to his settings.